### PR TITLE
[ACS-9736] Add new bean for intercepting current audit entries.

### PR DESCRIPTION
**Pull request by kmagdziarz**
This PR introduces a new audit record reporting mechanism for intercepting current audit entries. The implementation adds the ability to report audit data to audit storage in addition to the traditional database storage.

Key changes:

- Added new audit record classes and utilities for structured audit data handling
- Introduced configurable audit destination settings (database vs audit storage)
- Created audit record reporter interface and implementation

### CODE REVIEW

File: repository/src/main/java/org/alfresco/repo/audit/AuditComponent.java [hunk 1]
```diff
@@ -70,6 +70,13 @@
      */
     public void setUserAuditFilter(UserAuditFilter userAuditFilter);
 
+    /**
+     * @param auditRecordReporter
+     *            AuditRecordReporter
+     * @since 25.2
+     */
+    public void setAuditRecordReporter(AuditRecordReporter auditRecordReporter);
```
[the remark this answers is not in the document]
Comment (kmagdziarz) [reply repository/src/main/java/org/alfresco/repo/audit/AuditComponent.java:78]: I adjusted code to already existing one.

File: repository/src/main/java/org/alfresco/repo/audit/AuditComponentImpl.java [hunk 2]
```diff
@@ -743,8 +743,8 @@ public Map<String, Serializable> doWork() throws Exception
         {
             String root = value.getKey();
             int index = root.lastIndexOf("/");
-            Map<String, Serializable> argc = new HashMap<String, Serializable>(1);
-            argc.put(root.substring(index, root.length()).substring(1), value.getValue());
+            Map<String, Serializable> argc = new HashMap<>(1);
+            argc.put(root.substring(index).substring(1), value.getValue());
```
Comment (cezary-witkowski) [inline repository/src/main/java/org/alfresco/repo/audit/AuditComponentImpl.java:747]: Optional. if you want to improve the readibility for it even more, you could extract to variable with meaningful name. Example 
```
var meaningfulName = root.substring(index).substring(1);
argc.put(meaningfulName, value.getValue());
```

File: repository/src/main/java/org/alfresco/repo/audit/model/AuditModelRegistryImpl.java [hunk 3]
```diff
@@ -249,6 +252,26 @@ public boolean isAuditEnabled()
...
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isAuditingToAuditStorageEnabled()
+    {
+        String value = getProperty(AUDIT_PROPERTY_AUDIT_ENABLED + AUDITING_TO_AUDIT_STORAGE);
+        return value != null && value.equalsIgnoreCase("true");
+    }
```
Comment (cezary-witkowski) [inline repository/src/main/java/org/alfresco/repo/audit/model/AuditModelRegistryImpl.java:273]: isAuditingToDatabaseEnabled()
| value | result |
| - | - |
| null | true |
| "true" | true |
| "anything else" | false |

isAuditingToAuditStorageEnabled()
| value | result |
| - | - |
| null | false⚠️ |
| "true" | true |
| "anything else" | false |

Is this difference by design?

File: repository/src/main/java/org/alfresco/repo/audit/AuditRecordUtils.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+ */
+package org.alfresco.repo.audit;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.alfresco.service.cmr.repository.NodeRef;
+
+public class AuditRecordUtils
```
Comment (cezary-witkowski) [inline repository/src/main/java/org/alfresco/repo/audit/AuditRecordUtils.java]: final to discourage extending

File: repository/src/main/java/org/alfresco/repo/audit/AuditComponentImpl.java [hunk 5]
```diff
@@ -925,6 +930,28 @@ private Map<String, Serializable> generateData(Map<String, DataGenerator> genera
...
+     * Creates an AuditRecord from the provided audit data.
+     *
+     * @param auditData
+     *            the map containing audit data
+     * @param inTransaction
+     *            whether the record is created within a transaction
+     * @param applicationName
+     *            the name of the audit application
+     * @return a constructed AuditRecord instance
+     */
```
Comment (cezary-witkowski) [inline repository/src/main/java/org/alfresco/repo/audit/AuditComponentImpl.java]: To me this javadoc is just noice and no additional information that I get from reading method signature. Consider removing or adding some useful info.

File: repository/src/main/java/org/alfresco/repo/audit/AuditComponentImpl.java [hunk 6]
```diff
@@ -925,6 +930,28 @@ private Map<String, Serializable> generateData(Map<String, DataGenerator> genera
...
+    private AuditRecord createAuditRecord(Map<String, Serializable> auditData, boolean inTransaction, String username, Long entryId, String applicationName)
+    {
+        int rootSize = applicationName.length() + 2; // Root is constructed like this -> '/' + auditedApplicationName + '/'.
```
Comment (cezary-witkowski) [inline repository/src/main/java/org/alfresco/repo/audit/AuditComponentImpl.java:938]: Maybe we just pass applicationName and do this calculation somewhere closer to where this rootSize is needed?

File: repository/src/main/java/org/alfresco/repo/audit/AuditRecordUtils.java (hunk 4 above)
Comment (cezary-witkowski) [reply repository/src/main/java/org/alfresco/repo/audit/AuditRecordUtils.java]: BTW are we sure the functionality of this class will never need to be overridden or extended in any way? Maybe a bean is the way to go?

File: repository/src/test/java/org/alfresco/repo/audit/AuditRecordUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.junit.Test;
+
+import org.alfresco.service.cmr.repository.NodeRef;
+import org.alfresco.service.namespace.QName;
+
+public class AuditRecordUtilsTest
```
Comment (cezary-witkowski) [inline repository/src/test/java/org/alfresco/repo/audit/AuditRecordUtilsTest.java:43]: I would like to see some edge case coverage. What if testData is empty? What if keyRoot is weird e.g. "/l/" (lowercase L)?. Maybe the root can be nothing (empty string)?

**Review by cezary-witkowski — APPROVED** [review]
Nice code

File: repository/src/main/java/org/alfresco/repo/audit/model/AuditModelRegistryImpl.java (hunk 3 above)
Comment (kmagdziarz) [reply repository/src/main/java/org/alfresco/repo/audit/model/AuditModelRegistryImpl.java:273]: Yes. isAuditingToDatabaseEnabled() should be enabled by default to not alter original behavior.

File: repository/src/test/java/org/alfresco/repo/audit/AuditRecordUtilsTest.java (hunk 7 above)
Comment (kmagdziarz) [reply repository/src/test/java/org/alfresco/repo/audit/AuditRecordUtilsTest.java:43]: At this point in the AuditComponent, it should not be possible for root to be empty.
But I agree about more edge cases.

File: repository/src/main/java/org/alfresco/repo/audit/AuditComponentImpl.java (hunk 6 above)
Comment (kmagdziarz) [reply repository/src/main/java/org/alfresco/repo/audit/AuditComponentImpl.java:938]: I prefer to do it there - it may change if future how it will be calculated.

File: repository/src/main/java/org/alfresco/repo/audit/AuditRecordUtils.java (hunk 4 above)
Comment (kmagdziarz) [reply repository/src/main/java/org/alfresco/repo/audit/AuditRecordUtils.java]: No, it should not be overridden or extended anywhere. 
Perhaps it implementation should be move to ent repo.

File: repository/src/test/java/org/alfresco/repo/audit/AuditRecordUtilsTest.java (hunk 7 above)
Comment (kmagdziarz) [reply repository/src/test/java/org/alfresco/repo/audit/AuditRecordUtilsTest.java:43]: And by the time AuditRecordUtils.generateAuditRecordBuilder() is called, testData cannot be empty.
